File: python/leet_code/205.py
"""
This module provides a solution for determining if two strings are isomorphic.

Two strings s and t are isomorphic if the characters in s can be replaced to get t, with preserving the order of characters.
No two characters may map to the same character, but a character may map to itself.

This module requires Python 3.12 and uses a dictionary for mapping the characters.

Example Usage:
    >>> s = Solution()
    >>> s.isIsomorphic('egg', 'add')
    True
    >>> s.isIsomorphic('foo', 'bar')
    False
    >>> s.isIsomorphic('paper', 'title')
    True
"""


class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        """
        Determines if two strings are isomorphic.

        This function creates a dictionary to store the mapping of characters from string s to string t.
        It then iterates over the characters in the strings. If a character in s is already mapped,
        it checks if the mapping is correct. If a character in s is not mapped, it checks if the character in t is already mapped.

        Args:
            s (str): The first string to compare.
            t (str): The second string to compare.

        Returns:
            bool: True if the strings are isomorphic, False otherwise.

        Raises:
            ValueError: If the lengths of the strings are not equal.

        Doctest:
            >>> s = Solution()
            >>> s.isIsomorphic('egg', 'add')
            True
            >>> s.isIsomorphic('foo', 'bar')
            False
            >>> s.isIsomorphic('paper', 'title')
            True
        """
        # Check if the lengths of the strings are equal
        if len(s) != len(t):
            raise ValueError('The lengths of the strings must be equal.')

        # Create a dictionary to store the mapping of characters
        mapping = {}

        # Iterate over the characters in the strings
        for i in range(len(s)):
            # Check if the character in s is already mapped
            if s[i] in mapping:
                # Check if the mapping is correct
                if mapping[s[i]] != t[i]:
                    return False
            else:
                # Check if the character in t is already mapped
                if t[i] in mapping.values():
                    return False
                mapping[s[i]] = t[i]

        return True

File: go/leet_code/205.go
// Package leet_code provides solutions for LeetCode problems.
//
// This file provides a solution for the problem of determining if two strings are isomorphic.
// Two strings s and t are isomorphic if the characters in s can be replaced to get t.
// All occurrences of a character must be replaced with another character while preserving the order of characters.
// No two characters may map to the same character, but a character may map to itself.
package leet_code

// isIsomorphic determines if two strings are isomorphic.
//
// It accepts two parameters:
// - s: a string representing the first string.
// - t: a string representing the second string.
//
// The function returns a boolean indicating whether the two strings are isomorphic.
//
// The function creates two maps to track the mapping from s to t and from t to s.
// It iterates over the strings, checking and updating the mappings.
// If a mapping that violates the isomorphism is found, the function returns false.
//
// Error Handling:
// - If the lengths of s and t are not equal, the function returns false.
//
// Time complexity analysis:
// - Best-case: O(n), when the strings are isomorphic.
// - Worst-case: O(n), when the strings are not isomorphic.
// - Average-case: O(n), as we always have to iterate over each character in the strings.
func isIsomorphic(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}

	m1 := make(map[byte]byte)
	m2 := make(map[byte]byte)

	for i := 0; i < len(s); i++ {
		if _, ok := m1[s[i]]; !ok {
			m1[s[i]] = t[i]
		} else if m1[s[i]] != t[i] {
			return false
		}

		if _, ok := m2[t[i]]; !ok {
			m2[t[i]] = s[i]
		} else if m2[t[i]] != s[i] {
			return false
		}
	}

	return true
}
